### ISSUE TITLE
Warning on function values

### ISSUE THREAD

**Issue by nyarly** [samsalisbury/psyringe#4]
Recently very confusing:

A variadic function cannot be used as a constructor, and so gets implicitly treated as a value.

On the one hand: there should be no reason not to make every function available as a value.

On the other, functions that can only be values, when their return type is an expected injection type, should issue some kind of warning - maybe only in Test()

**Comment by samsalisbury** [comment]
> A variadic function cannot be used as a constructor, and so gets implicitly treated as a value.

That's right, because injecting into variadics would make psyringe more complex without adding much function as far as I can tell. You can use a slice instead though.

> On the one hand: there should be no reason not to make every function available as a value.

If you want to make a constructor type available as a value, you need to return it from another constructor type. This design is intended to make it really simple to inject concrete non-constructor values, which are assumed to be the most common. Since everything is a singleton, the semantics of making all constructors available as injection types would make Psyringe harder to understand. Should they still return singletons (i.e. we create a new function with the same signature, and only run the injected constructor once), or does that mean we run the constructor again each time? If we assume the former, then injecting the constructor itself would not be useful, if we assume the latter we are breaking the main assumption that everything is a singleton, which makes reasoning about Psyringe harder for users. Psyringe is purposefully as simple as possible, but is easy to wrap in something more advanced if that's what you need.

> On the other, functions that can only be values, when their return type is an expected injection type, should issue some kind of warning - maybe only in Test()

You mean functions which are not constructors? Hm, maybe that is slightly confusing. I would not like to add a warning in tests for something that may in fact be legitimate, then it's just noise. However, to address the confusion I would be happy to limit Psyringe to accepting _only_ constructors, and not concrete values into its Add and New methods. That would just mean that non-constructors need to be wrapped in a function, which may actually make things more readable anyway, as you would be forced to specify the injection type explicitly via the function declaration.
